### ISSUE TITLE
Model still training after freezing all layers. Problem with model architecture?

### ISSUE THREAD

**Issue by atharvavaidya14** [ultralytics/ultralytics#4226]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am doing transfer learning on 'yolov8x-pose.pt'. And with this [Ultralytics transfer learning tutorial](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/) as reference, I tried freezing all 22 layers of the network and trained it with 500 images as shown below

```
def freeze_layer(trainer, num_freeze):
    model = trainer.model
    #num_freeze = 20
    print(f"Freezing {num_freeze} layers")
    freeze = [f'model.{x}.' for x in range(num_freeze)] # layers to freeze
    for k, v in model.named_parameters():
        print(k)
        v.requires_grad = True # train all layers
        if any(x in k for x in freeze):
            #print(f'freezing {k}')
            v.requires_grad = False
    print(f"{num_freeze} layers are freezed.")
if __name__ == "__main__":
  # shortened code for simplicity
  model.add_callback("on_train_start", lambda trainer: freeze_layer(trainer, layers_to_freeze))
  model.train(
          data='data.yaml',
          rect=True, 
          epochs=10, 
          patience=0,
          imgsz=640, 
          batch=16, 
          close_mosaic=1, 
          save=True, 
          project=dir, 
          name=proj_name,
      )
```

This is the results.png file generated which shows clear improvement even after freezing all layers.

![results](https://github.com/ultralytics/ultralytics/assets/97112478/7a676417-f619-446c-bcc1-5d6956f01f85)

Is there any other way to freeze the layers or is this a bug? This issue is also faced by other people like in #3862 

I also have another question about the model's architecture.
When I printed out the layers with this `for k, v in model.named_parameters():
print(k)`, I noticed in the output shown below that some layer numbers are missing while some are repeated a lot. Do these numbers not correspond to actual layer numbers and are more of an indication of the type of layers? If not, why does the yaml file for the model still only specifies 22 layers and not all of these shown below.

```
model.0.conv.weight
model.0.bn.weight
model.0.bn.bias
model.1.conv.weight
model.1.bn.weight
model.1.bn.bias
model.2.cv1.conv.weight
model.2.cv1.bn.weight
model.2.cv1.bn.bias
model.2.cv2.conv.weight
model.2.cv2.bn.weight
model.2.cv2.bn.bias
model.2.m.0.cv1.conv.weight
model.2.m.0.cv1.bn.weight
model.2.m.0.cv1.bn.bias
model.2.m.0.cv2.conv.weight
model.2.m.0.cv2.bn.weight
model.2.m.0.cv2.bn.bias
model.2.m.1.cv1.conv.weight
model.2.m.1.cv1.bn.weight
model.2.m.1.cv1.bn.bias
model.2.m.1.cv2.conv.weight
model.2.m.1.cv2.bn.weight
model.2.m.1.cv2.bn.bias
model.2.m.2.cv1.conv.weight
model.2.m.2.cv1.bn.weight
model.2.m.2.cv1.bn.bias
model.2.m.2.cv2.conv.weight
model.2.m.2.cv2.bn.weight
model.2.m.2.cv2.bn.bias
model.3.conv.weight
model.3.bn.weight
model.3.bn.bias
model.4.cv1.conv.weight
model.4.cv1.bn.weight
model.4.cv1.bn.bias
model.4.cv2.conv.weight
model.4.cv2.bn.weight
model.4.cv2.bn.bias
model.4.m.0.cv1.conv.weight
model.4.m.0.cv1.bn.weight
model.4.m.0.cv1.bn.bias
model.4.m.0.cv2.conv.weight
model.4.m.0.cv2.bn.weight
model.4.m.0.cv2.bn.bias
model.4.m.1.cv1.conv.weight
model.4.m.1.cv1.bn.weight
model.4.m.1.cv1.bn.bias
model.4.m.1.cv2.conv.weight
model.4.m.1.cv2.bn.weight
model.4.m.1.cv2.bn.bias
model.4.m.2.cv1.conv.weight
model.4.m.2.cv1.bn.weight
model.4.m.2.cv1.bn.bias
model.4.m.2.cv2.conv.weight
model.4.m.2.cv2.bn.weight
model.4.m.2.cv2.bn.bias
model.4.m.3.cv1.conv.weight
model.4.m.3.cv1.bn.weight
model.4.m.3.cv1.bn.bias
model.4.m.3.cv2.conv.weight
model.4.m.3.cv2.bn.weight
model.4.m.3.cv2.bn.bias
model.4.m.4.cv1.conv.weight
model.4.m.4.cv1.bn.weight
model.4.m.4.cv1.bn.bias
model.4.m.4.cv2.conv.weight
model.4.m.4.cv2.bn.weight
model.4.m.4.cv2.bn.bias
model.4.m.5.cv1.conv.weight
model.4.m.5.cv1.bn.weight
model.4.m.5.cv1.bn.bias
model.4.m.5.cv2.conv.weight
model.4.m.5.cv2.bn.weight
model.4.m.5.cv2.bn.bias
model.5.conv.weight
model.5.bn.weight
model.5.bn.bias
model.6.cv1.conv.weight
model.6.cv1.bn.weight
model.6.cv1.bn.bias
model.6.cv2.conv.weight
model.6.cv2.bn.weight
model.6.cv2.bn.bias
model.6.m.0.cv1.conv.weight
model.6.m.0.cv1.bn.weight
model.6.m.0.cv1.bn.bias
model.6.m.0.cv2.conv.weight
model.6.m.0.cv2.bn.weight
model.6.m.0.cv2.bn.bias
model.6.m.1.cv1.conv.weight
model.6.m.1.cv1.bn.weight
model.6.m.1.cv1.bn.bias
model.6.m.1.cv2.conv.weight
model.6.m.1.cv2.bn.weight
model.6.m.1.cv2.bn.bias
model.6.m.2.cv1.conv.weight
model.6.m.2.cv1.bn.weight
model.6.m.2.cv1.bn.bias
model.6.m.2.cv2.conv.weight
model.6.m.2.cv2.bn.weight
model.6.m.2.cv2.bn.bias
model.6.m.3.cv1.conv.weight
model.6.m.3.cv1.bn.weight
model.6.m.3.cv1.bn.bias
model.6.m.3.cv2.conv.weight
model.6.m.3.cv2.bn.weight
model.6.m.3.cv2.bn.bias
model.6.m.4.cv1.conv.weight
model.6.m.4.cv1.bn.weight
model.6.m.4.cv1.bn.bias
model.6.m.4.cv2.conv.weight
model.6.m.4.cv2.bn.weight
model.6.m.4.cv2.bn.bias
model.6.m.5.cv1.conv.weight
model.6.m.5.cv1.bn.weight
model.6.m.5.cv1.bn.bias
model.6.m.5.cv2.conv.weight
model.6.m.5.cv2.bn.weight
model.6.m.5.cv2.bn.bias
model.7.conv.weight
model.7.bn.weight
model.7.bn.bias
model.8.cv1.conv.weight
model.8.cv1.bn.weight
model.8.cv1.bn.bias
model.8.cv2.conv.weight
model.8.cv2.bn.weight
model.8.cv2.bn.bias
model.8.m.0.cv1.conv.weight
model.8.m.0.cv1.bn.weight
model.8.m.0.cv1.bn.bias
model.8.m.0.cv2.conv.weight
model.8.m.0.cv2.bn.weight
model.8.m.0.cv2.bn.bias
model.8.m.1.cv1.conv.weight
model.8.m.1.cv1.bn.weight
model.8.m.1.cv1.bn.bias
model.8.m.1.cv2.conv.weight
model.8.m.1.cv2.bn.weight
model.8.m.1.cv2.bn.bias
model.8.m.2.cv1.conv.weight
model.8.m.2.cv1.bn.weight
model.8.m.2.cv1.bn.bias
model.8.m.2.cv2.conv.weight
model.8.m.2.cv2.bn.weight
model.8.m.2.cv2.bn.bias
model.9.cv1.conv.weight
model.9.cv1.bn.weight
model.9.cv1.bn.bias
model.9.cv2.conv.weight
model.9.cv2.bn.weight
model.9.cv2.bn.bias
model.12.cv1.conv.weight
model.12.cv1.bn.weight
model.12.cv1.bn.bias
model.12.cv2.conv.weight
model.12.cv2.bn.weight
model.12.cv2.bn.bias
model.12.m.0.cv1.conv.weight
model.12.m.0.cv1.bn.weight
model.12.m.0.cv1.bn.bias
model.12.m.0.cv2.conv.weight
model.12.m.0.cv2.bn.weight
model.12.m.0.cv2.bn.bias
model.12.m.1.cv1.conv.weight
model.12.m.1.cv1.bn.weight
model.12.m.1.cv1.bn.bias
model.12.m.1.cv2.conv.weight
model.12.m.1.cv2.bn.weight
model.12.m.1.cv2.bn.bias
model.12.m.2.cv1.conv.weight
model.12.m.2.cv1.bn.weight
model.12.m.2.cv1.bn.bias
model.12.m.2.cv2.conv.weight
model.12.m.2.cv2.bn.weight
model.12.m.2.cv2.bn.bias
model.15.cv1.conv.weight
model.15.cv1.bn.weight
model.15.cv1.bn.bias
model.15.cv2.conv.weight
model.15.cv2.bn.weight
model.15.cv2.bn.bias
model.15.m.0.cv1.conv.weight
model.15.m.0.cv1.bn.weight
model.15.m.0.cv1.bn.bias
model.15.m.0.cv2.conv.weight
model.15.m.0.cv2.bn.weight
model.15.m.0.cv2.bn.bias
model.15.m.1.cv1.conv.weight
model.15.m.1.cv1.bn.weight
model.15.m.1.cv1.bn.bias
model.15.m.1.cv2.conv.weight
model.15.m.1.cv2.bn.weight
model.15.m.1.cv2.bn.bias
model.15.m.2.cv1.conv.weight
model.15.m.2.cv1.bn.weight
model.15.m.2.cv1.bn.bias
model.15.m.2.cv2.conv.weight
model.15.m.2.cv2.bn.weight
model.15.m.2.cv2.bn.bias
model.16.conv.weight
model.16.bn.weight
model.16.bn.bias
model.18.cv1.conv.weight
model.18.cv1.bn.weight
model.18.cv1.bn.bias
model.18.cv2.conv.weight
model.18.cv2.bn.weight
model.18.cv2.bn.bias
model.18.m.0.cv1.conv.weight
model.18.m.0.cv1.bn.weight
model.18.m.0.cv1.bn.bias
model.18.m.0.cv2.conv.weight
model.18.m.0.cv2.bn.weight
model.18.m.0.cv2.bn.bias
model.18.m.1.cv1.conv.weight
model.18.m.1.cv1.bn.weight
model.18.m.1.cv1.bn.bias
model.18.m.1.cv2.conv.weight
model.18.m.1.cv2.bn.weight
model.18.m.1.cv2.bn.bias
model.18.m.2.cv1.conv.weight
model.18.m.2.cv1.bn.weight
model.18.m.2.cv1.bn.bias
model.18.m.2.cv2.conv.weight
model.18.m.2.cv2.bn.weight
model.18.m.2.cv2.bn.bias
model.19.conv.weight
model.19.bn.weight
model.19.bn.bias
model.21.cv1.conv.weight
model.21.cv1.bn.weight
model.21.cv1.bn.bias
model.21.cv2.conv.weight
model.21.cv2.bn.weight
model.21.cv2.bn.bias
model.21.m.0.cv1.conv.weight
model.21.m.0.cv1.bn.weight
model.21.m.0.cv1.bn.bias
model.21.m.0.cv2.conv.weight
model.21.m.0.cv2.bn.weight
model.21.m.0.cv2.bn.bias
model.21.m.1.cv1.conv.weight
model.21.m.1.cv1.bn.weight
model.21.m.1.cv1.bn.bias
model.21.m.1.cv2.conv.weight
model.21.m.1.cv2.bn.weight
model.21.m.1.cv2.bn.bias
model.21.m.2.cv1.conv.weight
model.21.m.2.cv1.bn.weight
model.21.m.2.cv1.bn.bias
model.21.m.2.cv2.conv.weight
model.21.m.2.cv2.bn.weight
model.21.m.2.cv2.bn.bias
model.22.cv2.0.0.conv.weight
model.22.cv2.0.0.bn.weight
model.22.cv2.0.0.bn.bias
model.22.cv2.0.1.conv.weight
model.22.cv2.0.1.bn.weight
model.22.cv2.0.1.bn.bias
model.22.cv2.0.2.weight
model.22.cv2.0.2.bias
model.22.cv2.1.0.conv.weight
model.22.cv2.1.0.bn.weight
model.22.cv2.1.0.bn.bias
model.22.cv2.1.1.conv.weight
model.22.cv2.1.1.bn.weight
model.22.cv2.1.1.bn.bias
model.22.cv2.1.2.weight
model.22.cv2.1.2.bias
model.22.cv2.2.0.conv.weight
model.22.cv2.2.0.bn.weight
model.22.cv2.2.0.bn.bias
model.22.cv2.2.1.conv.weight
model.22.cv2.2.1.bn.weight
model.22.cv2.2.1.bn.bias
model.22.cv2.2.2.weight
model.22.cv2.2.2.bias
model.22.cv3.0.0.conv.weight
model.22.cv3.0.0.bn.weight
model.22.cv3.0.0.bn.bias
model.22.cv3.0.1.conv.weight
model.22.cv3.0.1.bn.weight
model.22.cv3.0.1.bn.bias
model.22.cv3.0.2.weight
model.22.cv3.0.2.bias
model.22.cv3.1.0.conv.weight
model.22.cv3.1.0.bn.weight
model.22.cv3.1.0.bn.bias
model.22.cv3.1.1.conv.weight
model.22.cv3.1.1.bn.weight
model.22.cv3.1.1.bn.bias
model.22.cv3.1.2.weight
model.22.cv3.1.2.bias
model.22.cv3.2.0.conv.weight
model.22.cv3.2.0.bn.weight
model.22.cv3.2.0.bn.bias
model.22.cv3.2.1.conv.weight
model.22.cv3.2.1.bn.weight
model.22.cv3.2.1.bn.bias
model.22.cv3.2.2.weight
model.22.cv3.2.2.bias
model.22.dfl.conv.weight
model.22.cv4.0.0.conv.weight
model.22.cv4.0.0.bn.weight
model.22.cv4.0.0.bn.bias
model.22.cv4.0.1.conv.weight
model.22.cv4.0.1.bn.weight
model.22.cv4.0.1.bn.bias
model.22.cv4.0.2.weight
model.22.cv4.0.2.bias
model.22.cv4.1.0.conv.weight
model.22.cv4.1.0.bn.weight
model.22.cv4.1.0.bn.bias
model.22.cv4.1.1.conv.weight
model.22.cv4.1.1.bn.weight
model.22.cv4.1.1.bn.bias
model.22.cv4.1.2.weight
model.22.cv4.1.2.bias
model.22.cv4.2.0.conv.weight
model.22.cv4.2.0.bn.weight
model.22.cv4.2.0.bn.bias
model.22.cv4.2.1.conv.weight
model.22.cv4.2.1.bn.weight
model.22.cv4.2.1.bn.bias
model.22.cv4.2.2.weight
model.22.cv4.2.2.bias

```

**Comment by Kayzwer** [comment]
@atharvavaidya14 dont forget this set warmup_epochs to 0, cuz lr is high during warmup_epochs.

**Comment by glenn-jocher** [comment]
Sure @atharvavaidya14, you are correct. Setting warmup_epochs to 0 is a good tip for your training process as it allows your learning rate to begin from a high value. Warmup_epochs is used to gradually increase the learning rate from a lower value to the defined learning rate over a certain number of initial epochs to prevent drastic parameter updates. If you start from a lower learning rate, this could potentially enable more refined feature extraction during the initial epochs, but could also slow down initial learning. It's always a balance between model accuracy and training speed that you have to consider in your specific use case.

**Comment by atharvavaidya14** [comment]
@glenn-jocher @Kayzwer Thanks for the suggestion. I tried that but the result was still the same. My main problem is that the model is getting trained even when I freeze all 22 layers in the model. This should not happen. And since I am experimenting with freezing different number of layers and training, I cannot trust that it actually freezes the layers.

**Comment by atharvavaidya14** [comment]
Closing this issue because I found the mistake and it was a problem with my implementation

**Comment by glenn-jocher** [comment]
@atharvavaidya14 hello, glad to hear that you were able to find and fix the problem in your implementation. Please feel free to ask or share if you encounter any other challenges in future. Remember that contributing to the discussion can help others who may experience similar issues. Good luck with your continuing work with YOLOv8!

**Comment by Y-T-G** [comment]
@atharvavaidya14 What was the issue?

**Comment by ducknificient** [comment]
i'm facing the same problem

**Comment by glenn-jocher** [comment]
@ducknificient hey, I discovered the issue was in how I managed the layer freezing within my implementation. Make sure you're correctly iterating over your model's layers and setting `requires_grad` to `False` for the layers you intend to freeze. Here's a brief example for clarity:

```python
for name, param in model.named_parameters():
    if name.startswith('model.0.'):  # Example condition to identify layers to freeze
        param.requires_grad = False
```

Check your conditions for freezing the layers to ensure they align with your model's architecture. If you've followed a guide, revisit it to ensure every detail matches up. Hope this helps! 🛠️
